src/dbnode/storage/index/block.go

```diff
@@ -102,6 +106,61 @@ func (s blockState) String() string {
 	return "unknown"
 }
 
+// LimitBlocksQueried enforces a max blocks query limit within a recency window.
+func LimitBlocksQueried(max int, within time.Duration) chan struct{} {
+	recentlyQueried = &queryRecencyWindow{
+		length:         within,
+		maxBlocks:      max,
+		recentBlocks:   atomic.NewInt64(0),
+		previousBlocks: 0,
+		stopCh:         make(chan struct{}),
+	}
+	recentlyQueried.Start()
+
+	// Give caller handle to stop the background flushing.
+	return recentlyQueried.stopCh
+}
+
+// For tracking query stats in past X duration such as blocks queried.
+type queryRecencyWindow struct {
+	length    time.Duration
+	maxBlocks int
+
+	recentBlocks   *atomic.Int64
+	previousBlocks int64
+	stopCh         chan struct{}
+}
+
+func (w queryRecencyWindow) Start() {
+	ticker := time.NewTicker(w.length)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			// Clear older active blocks every X duration.
+			w.recentBlocks.Sub(w.previousBlocks)
+			w.previousBlocks = w.recentBlocks.Load()
+		case <-w.stopCh:
+			return
+		}
+	}
+}
+
+func (w *queryRecencyWindow) AddWithinLimit(blocks int) bool {
+	if w == nil {
+		return true
+	}
+
+	inc := int64(blocks)
+	w.recentBlocks.Add(inc)
+	val := int(w.recentBlocks.Load())
```

Comment:
You can use the value of `.Add(inc)` here instead of loading which will reduce the amount of atomic instructions required for each statement.

Reply:
Nice didn't realize `Add(...)` returned the result too